Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -38,59 +43,222 @@ open scoped Matrix
+
+end sqrt
+
+end PosSemidef
+
 @[simp]
 theorem posSemidef_submatrix_equiv {M : Matrix n n R} (e : m ≃ n) :
     (M.submatrix e e).PosSemidef ↔ M.PosSemidef :=
   ⟨fun h => by simpa using h.submatrix e.symm, fun h => h.submatrix _⟩
 #align matrix.pos_semidef_submatrix_equiv Matrix.posSemidef_submatrix_equiv
 
 /-- The conjugate transpose of a matrix mulitplied by the matrix is positive semidefinite -/
-theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : Matrix.PosSemidef (Aᴴ * A) := by
+theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : PosSemidef (Aᴴ * A) := by
   refine ⟨isHermitian_transpose_mul_self _, fun x => ?_⟩
   rw [← mulVec_mulVec, dotProduct_mulVec, vecMul_conjTranspose, star_star]
   exact Finset.sum_nonneg fun i _ => star_mul_self_nonneg _
 
 /-- A matrix multiplied by its conjugate transpose is positive semidefinite -/
-theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : Matrix.PosSemidef (A * Aᴴ) :=
+theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : PosSemidef (A * Aᴴ) :=
   by simpa only [conjTranspose_conjTranspose] using posSemidef_conjTranspose_mul_self Aᴴ
 
-/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
-lemma PosSemidef.eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
-    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
-  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
-
 lemma eigenvalues_conjTranspose_mul_self_nonneg (A : Matrix m n 𝕜) [DecidableEq n] (i : n) :
     0 ≤ (isHermitian_transpose_mul_self A).eigenvalues i :=
-  (Matrix.posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
+  (posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
 
 lemma eigenvalues_self_mul_conjTranspose_nonneg (A : Matrix m n 𝕜) [DecidableEq m] (i : m) :
     0 ≤ (isHermitian_mul_conjTranspose_self A).eigenvalues i :=
-  (Matrix.posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+  (posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+
+/-- A matrix is positive semidefinite if and only if it has the form `Bᴴ * B` for some `B`. -/
+lemma posSemidef_iff_eq_transpose_mul_self [DecidableEq n] {A : Matrix n n 𝕜} :
+    PosSemidef A ↔ ∃ (B : Matrix n n 𝕜), A = Bᴴ * B := by
+  refine ⟨fun hA ↦ ⟨hA.sqrt, ?_⟩, fun ⟨B, hB⟩ ↦ (hB ▸ posSemidef_conjTranspose_mul_self B)⟩
+  simp_rw [← PosSemidef.sq_sqrt hA, pow_two]
+  rw [hA.posSemidef_sqrt.1]
+
+/-- For `A` positive semidefinite, we have `x⋆ A x = 0` iff `A x = 0`. -/
+theorem PosSemidef.dotProduct_mulVec_zero_iff [DecidableEq n]
+    {A : Matrix n n 𝕜} (hA : PosSemidef A) (x : n → 𝕜) :
+    star x ⬝ᵥ mulVec A x = 0 ↔ mulVec A x = 0 := by
+  constructor
+  · obtain ⟨B, rfl⟩ := posSemidef_iff_eq_transpose_mul_self.mp hA
+    rw [← Matrix.mulVec_mulVec, dotProduct_mulVec,
+      vecMul_conjTranspose, star_star, dotProduct_star_self_eq_zero]
+    intro h0
+    rw [h0, mulVec_zero]
+  · intro h0
+    rw [h0, dotProduct_zero]
```

Comment:
Maybe this instead:
```suggestion
/-- For `A` positive semidefinite, we have `x⋆ A x = 0` iff `√A x = 0` iff `A x = 0`. -/
theorem PosSemidef.mulVec_tfae [DecidableEq n]
    {A : Matrix n n 𝕜} (hA : PosSemidef A) (x : n → 𝕜) :
    List.TFAE [star x ⬝ᵥ mulVec A x = 0, mulVec hA.sqrt x = 0, mulVec A x = 0] := by
  tfae_have 1 → 2
  · intro h
    rwa [← hA.sq_sqrt, sq, ← mulVec_mulVec, dotProduct_mulVec, ← hA.posSemidef_sqrt.1.eq,
      vecMul_conjTranspose, star_star, hA.posSemidef_sqrt.1.eq, dotProduct_star_self_eq_zero] at h
  tfae_have 2 → 3
  · intro h
    rw [← hA.sq_sqrt, sq, ← mulVec_mulVec, h, mulVec_zero]
  tfae_have 3 → 1
  · intro h
    simp [h]
  tfae_finish
```

Reply:
I'm not convinced by this: the "middle" condition in the TFAE list seems like an artefact of the proof, not something which deserves to go in the statement. I don't feel that adding it outweighs the usability hit that comes with formulating it as a TFAE. 

I also strongly suspect that the equivalence "`star x ⬝ᵥ mulVec A x = 0` iff `mulVec A x = 0` holds much more generally. It certainly holds for positive-semidef symmetric matrices over any Archimedean linearly ordered field: any such field embeds in the reals (although I think mathlib doesn't know this yet), and the two conditions are both preserved by scalar extension, so the general case follows from the real case. Of course, the vast majority of such fields aren't closed under square roots so the middle condition doesn't even make sense for them. So, with an eye to future generalisations, I'd rather we didn't build square roots into the statement.